### ISSUE TITLE
I get a 502 error when updating a page property on database with a large number of pages (3000 over)

### ISSUE THREAD

**Issue by 75asa** [makenotion/notion-sdk-js#172]
Report bugs here only for the Node JavaScript library.

If you're having problems using Notion's API, or have any other feedback about the API including feature requests for the JavaScript library, please email support at developers@notion.com.

**Describe the bug**
We are currently moving a lot of imported data into the database and attaching properties etc.
The update process itself is done synchronously, and works fine until the middle of the process, but when it reaches a certain number, I get a 502 error.
Is there any workaround for this?
my code is [here](https://github.com/75asa/kibela-to-notion)

**To Reproduce**
Node version: v16.4.0
Notion JS library version: ^0.2.3

Steps to reproduce the behavior: update page props on database includes over 3000 pages.

**Expected behavior**
The update must be successful.

**Screenshots**
The error code is as follows
```
@notionhq/client warn: request fail {
  code: 'notionhq_client_response_error',
  message: 'Request to Notion API failed with status: 502'
}
```

**Additional context**
The specific code is below

https://github.com/75asa/kibela-to-notion/blob/91639d1350a8fd608c626a7bbe24a937996a001e/src/repository/NotionRepository.ts#L56

**Comment by GorvGoyl** [comment]
faced the same issue even though I wasn't updating in bulk.

**Comment by 75asa** [comment]
Hi @GorvGoyl
A few days ago, I had got this message on Notion SDK community. by @97agupta 

> Hey y'all! We are actively investigating these 502's but could use some help getting access to one of your databases that's returning this error. If you can help please DM me!

I've sended a detail on this issue. 
If you'd like, you might want to DM them too.
Thax :)

**Comment by rhart92** [comment]
Closing due to inactivity.

**Comment by loong** [comment]
experiencing the same issue on even a smaller number (300 pages)

**Comment by provishalk** [comment]
> experiencing the same issue on even a smaller number (300 pages)

Facing the same issue, I just wanted to know if you are using the npm package for this.

**Comment by loong** [comment]
@provishalk I'm actually using this with deno importing directly from deno.land!

```javascript
import { Client } from 'https://deno.land/x/notion_sdk/src/mod.ts'

const notion = new Client({ auth: Deno.env.get('NOTION_KEY') }) 

// ...

// this gets called in a loop
  await notion.pages.create({                                                                                                                                                                                                               
    parent: {                                                                                                                                                                                                                                      
      type: 'database_id',                                                                                                                                                                                                                         
      database_id: dbId,                                                                                                                                                                                                                           
    },                                                                                                                                                                                                                                             
    properties: {                                                                                                                                                                                                                                  
      label: {                                                                                                                                                                                                                                     
        title: [                                                                                                                                                                                                                                   
          {                                                                                                                                                                                                                                        
            text: {                                                                                                                                                                                                                                
              content: prop.label,                                                                                                                                                                                                                 
            },                                                                                                                                                                                                                                     
          },                                                                                                                                                                                                                                       
        ],                                                                                                                                                                                                                                         
      },   
      // ....
    },                                                                                                                                                                                                                                             
  })  
```

**Comment by jeremie-val** [comment]
Same issue here @rhart92, can we have an update ?

**Comment by anand-yadav55** [comment]
@75asa , @rhart92 any fixes for this? I am getting this issue, it arises and vanishes by itself.

**Comment by toficofi** [comment]
Also experiencing this issue

**Comment by MSBIZ** [comment]
same problem

**Comment by Kimblis** [comment]
Getting this error on page retrieval as well. Any news on this?

**Comment by DEFCESCO** [comment]
Hello, I, too, have been running into this issue. I am using a workaround by handling the error and retrying the request with an exponential backoff. The request works after the second or third time. I have not seen a failure since. 

I am sending thousands of requests, and it does not appear to be a rate-limiting issue. 

I also noticed Cloudflare had issues yesterday, [which are resolved today](https://cloudflarestatus.com/). This appears to be an issue with Notion's servers. 

As a premium subscriber to Notion, this was a hassle troubleshooting my newly implemented code.

Here is some reference code to retry a request:
```
def create_table_row_with_retry(table_block_id, row_data, max_retries=3):
    retries = 0
    while retries < max_retries:
        try:
            create_table_row(table_block_id, row_data)
            break  # Success, exit the loop
        except Exception as e:
            print(f"Error creating table row: {e}")
            retries += 1
            if retries < max_retries:
                # Exponential backoff: Wait before retrying
                wait_time = 2 ** retries
                print(f"Retrying in {wait_time} seconds...")
                time.sleep(wait_time)
            else:
                print("Max retries reached. Unable to create table row.")
```

**Comment by changingwang** [comment]
same problem.

raise HTTPResponseError(error.response)
(https://github.com/changingwang/weread_to_notion/actions/runs/6294624369/job/17086719823#step:5:33)
notion_client.errors.HTTPResponseError: Request to Notion API failed with status: 502
(https://github.com/changingwang/weread_to_notion/actions/runs/6294624369/job/17086719823#step:5:34)
Error: Process completed with exit code 1.

**Comment by cd-slash** [comment]
I'm getting this repeatedly on a relatively small database (low hundreds of rows) and it seems to be happening rather randomly. It's happening both in the SDK and when hitting the URL from Thunder Client.

When I get a `502 Bad Gateway` response, trying exactly the same query a few seconds later results in a successful response so it's really unclear why this is happening.

Suggest this should be reopened, though since it's happening with responses from the URL also, seems it's perhaps a wider API issue.

**Comment by DineshCodeFlow** [comment]
The issue should be reopened. is happening for me too!

**Comment by finkrer** [comment]
@rhart92 it seems like about 3 days ago something happened and a lot of people are getting 502 errors now, me included.

**Comment by Wouter125** [comment]
Yup, issues are back again. Can confirm.

**Comment by TanyaChichkan** [comment]
Same for me

**Comment by develicejus** [comment]
Same for me

**Comment by donaghhorgan** [comment]
Also seeing this sporadically with the API

**Comment by eacarras** [comment]
Same here. Trying retrieve data with the npm package and the url is returning 502 or timeout.

**Comment by victoralvelais** [comment]
Been getting this error bad since the weekend and now I have to prioritize an error handling solution.

**Comment by juanicortesl** [comment]
Seeing this error appear more often since last week.

**Comment by AlAgilly** [comment]
Hey, this issue has been reoccurring for me an increasing amount for the past few days.
DB has only 16 entries, and my updates work about 20% of the time; always throwing 502 error somewhere near the middle of the process.

**Comment by qkg982** [comment]
I also have the same problem with the response [502] from the API when fetching pages. This has worked flawlessly back in time, but last week I started getting the [502] response on random pages when fetching them via the API and the problem is still there.

**Comment by britonmarketing** [comment]
getting totally same, pls re-open!

**Comment by henri9813** [comment]
Notion has an outage ?

I randomly got 502 errors on getting pages thought the API

**Comment by sebastien-moha** [comment]
Same here.. I build a python ETL to extract data from notion database and load them into Postgresql. All was working fine but now, I just get 502 error randomly...

**Comment by PercussiveElbow** [comment]
Same issue. This has been going on for months in some capacity but even significantly reducing queries and adding large sleep timeouts doesn't work. The notion API really isn't production ready when issues like this are widespread.

**Comment by mclzion** [comment]
About 5 days ago we started getting 502 errors.
I guess it's related to Notion's API, but really not sure.
If there're tips for getting more verbose output I'd love to provide more context!

**Comment by Smidgens** [comment]
I'm building a static blog with Next.js and just this week I started seeing sporadic 502 errors when fetching page content. I'm currently rethinking my build strategy to work around it.

**Comment by accounts01** [comment]
+1. I'm experiencing this as well.

**Comment by dotansimha** [comment]
Same here, in the past week we had a lot of 500 and 502.

**Comment by DineshCodeFlow** [comment]
Tired seing this:-

@notionhq/client warn: request fail {
  code: 'notionhq_client_response_error',
  message: 'Request to Notion API failed with status: 502'
}

**Comment by cd-slash** [comment]
It's baffling how the Notion Status page shows all systems are working normally - based on the number of contributors in this thread, that's clearly not the case.

I've worked around this for now by just retrying failed requests (they generally work on the 2nd attempt and I haven't seen any fail after 3 attempts) but I'm focusing my attention on moving off the platform rather than spending more time troubleshooting.

**Comment by bureaubox** [comment]
502s... 502s everywhere!

**Comment by llccrr** [comment]
up, same here

**Comment by Ryan-McBride** [comment]
I'm getting similar behavior on batch deletes. I thought maybe I was getting rate limited (even though that should be a 429) but that doesn't seem to be the case. This has pretty severely hurt my workflow.

**Comment by tb-lokalguiden** [comment]
Same here

**Comment by sinandmrc7** [comment]
same here

**Comment by devArno88** [comment]
I'm not sure how useful this answer will be here for `notion-sdk-js`, but I thought I'd share just in case. 

tl;dr - solved my intermittent 502s (python) with pagination and retries (and ofc some extra logging)

I'm working with [notion-sdk-py](https://github.com/ramnes/notion-sdk-py) and after a few days of successful experimentation, today I was hit by a bunch of random intermittent 502 errors with no further context (very frustrating), which eventually led me to this issue. 

Like many on this thread I first thought this was a request limit issue, but the [Requests Limit](https://developers.notion.com/reference/request-limits) reference states that `Rate-limited requests will return a "rate_limited" error code (HTTP response status 429)`. Guess not then!

After a bit more digging/logging I noticed my `RowParser` module was only receiving 100 rows at a time, which is obviously an issue when most of your databases are 500+ rows. And as expected, the Notion API [pagination reference](https://developers.notion.com/reference/pagination) confirmed the limit, so I decided to try and implement a solution using `start_cursor`. I managed to bypass the pagination limit and retrieve the all database rows. Hooray! Buuut I was **_still_** seeing an occasional 502. 🙄

The only viable way forward that made sense from here was a retry mechanism, or "exponential backoff" (i.e. keep retrying the request with a progressively increasing delay, just so Notion doesn't throw a tantrum). After some prompt magic and half a pack of digestive biscuits, I found the following implementation to work pretty well for my project requirements, with no 502s in sight:

```python
import os
import json
from typing import List, Dict, Any, Optional
from notion_client import Client
from ..base import DataNotFoundError, RepositoryError
import logging
import time

logger = logging.getLogger(__file__)

class NotionRepository:

    def __init__(self, client: Client):
        self.client = client

    def _get_db_rows(self, dbuid: str) -> List[Dict[str, Any]]:
        """Fetch rows from a Notion database with retries, pagination, and logging."""
        try:
            if dbuid not in self._db_map:
                logger.error(f"No database mapping found for ID: {dbuid}")
                raise DataNotFoundError(f"No database mapping found for ID: {dbuid}")
            
            logger.info(f"Starting query for database {dbuid}...")
            all_results = []
            next_cursor = None
            attempt = 0

            while True:
                try:
                    logger.info(f"Attempt {attempt + 1}: Querying with start_cursor {next_cursor}")
                    response = self._query_with_retries(self._db_map[dbuid], start_cursor=next_cursor)
                    results = response.get("results", [])
                    all_results.extend(results)
                    
                    next_cursor = response.get("next_cursor")
                    if not response.get("has_more", False):
                        logger.info(f"Query completed for database {dbuid}, no more pages.")
                        break

                    logger.info(f"More pages to query. Next cursor: {next_cursor}")
                    time.sleep(0.1)  # Optional rate-limiting delay

                except Exception as query_error:
                    logger.error(f"Error while querying database {dbuid}: {query_error}")
                    raise

                attempt += 1
            
            logger.info(f"Successfully retrieved {len(all_results)} rows from database {dbuid}")
            return all_results

        except Exception as e:
            logger.error(f"Unexpected error while querying database ID {dbuid}: {e}")
            raise RepositoryError(f"Unexpected error while querying database ID {dbuid}: {e}") from e


    def _query_with_retries(self, database_id: str, start_cursor: Optional[str] = None, retries: int = 5) -> dict:
        """Query the Notion API with retry mechanism for transient errors"""
        attempt = 0
        delay = 1  # Initial delay in seconds

        while attempt < retries:
            try:
                logger.info(f"Querying database {database_id} with start_cursor {start_cursor} (attempt {attempt + 1})")
                # Perform the query
                response = self.client.databases.query(database_id, start_cursor=start_cursor)
                logger.info(f"Query successful for database {database_id}")
                return response
            
            except Exception as e:
                logger.error(f"Error querying database {database_id} on attempt {attempt + 1}: {e}")
                if "502" in str(e):  # Retry on 502 error
                    attempt += 1
                    logger.info(f"Retrying in {delay} seconds...")
                    time.sleep(delay)
                    delay *= 2  # Exponential backoff
                    continue
                # Log and raise for other errors
                logger.error(f"Non-transient error: {e}")
                raise

        logger.error(f"Failed to query database {database_id} after {retries} retries.")
        raise Exception(f"Failed to query database {database_id} after {retries} retries.")
```

This solution is giving me consistent 200s, with logs showing:

```bash
INFO:my_cool_project/core/repositories/notion_repository.py:Starting query for database MY_NOTION_DB_ID...
INFO:my_cool_project/core/repositories/notion_repository.py:Attempt 1: Querying with start_cursor None
INFO:my_cool_project/core/repositories/notion_repository.py:Querying database MY_NOTION_DB_ID with start_cursor None (attempt 1)
INFO:httpx:HTTP Request: POST https://api.notion.com/v1/databases/MY_NOTION_DB_ID/query "HTTP/1.1 200 OK"
INFO:my_cool_project/core/repositories/notion_repository.py:Query successful for database MY_NOTION_DB_ID
INFO:my_cool_project/core/repositories/notion_repository.py:More pages to query. Next cursor: NEXT_CURSOR_ID
INFO:my_cool_project/core/repositories/notion_repository.py:Attempt 2: Querying with start_cursor NEXT_CURSOR_ID
INFO:my_cool_project/core/repositories/notion_repository.py:Querying database MY_NOTION_DB_ID with start_cursor NEXT_CURSOR_ID (attempt 1)
INFO:httpx:HTTP Request: POST https://api.notion.com/v1/databases/MY_NOTION_DB_ID/query "HTTP/1.1 200 OK"
INFO:my_cool_project/core/repositories/notion_repository.py:Query successful for database MY_NOTION_DB_ID
INFO:my_cool_project/core/repositories/notion_repository.py:Query completed for database MY_NOTION_DB_ID, no more pages.
INFO:my_cool_project/core/repositories/notion_repository.py:Successfully retrieved 128 rows from database MY_NOTION_DB_ID
```
(with some values replaced for obvious reasons)

Not the prettiest solution, but will have to do for the time being as the 502s are too disruptive to my workflow. 

I obviously can't confirm whether this will work for `notion-sdk-js`, but it might be somewhere to start (as it seems there is a lot of unanswered frustration in this thread).

Hope this helps someone - good luck!
